src/main/kotlin/org/rust/cargo/project/workspace/FeatureGraph.kt

```diff
@@ -0,0 +1,125 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.cargo.project.workspace
+
+import org.rust.lang.utils.Node
+import org.rust.lang.utils.PresentableGraph
+import org.rust.lang.utils.PresentableNodeData
+
+enum class FeatureState {
+    Enabled,
+    Disabled;
+
+    fun toBoolean(): Boolean = when (this) {
+        Enabled -> true
+        Disabled -> false
+    }
+}
+
+typealias FeatureNode = Node<PackageFeature, Unit>
+
+data class PackageFeature(val pkg: CargoWorkspace.Package, val name: String) : PresentableNodeData {
+    override val text: String
+        get() = "$pkg.$name"
+}
+
+typealias WorkspaceFeaturesGraph = PresentableGraph<PackageFeature, Unit>
+
+class FeatureGraph private constructor(
+    private val graph: WorkspaceFeaturesGraph,
+    private val featureToNode: Map<PackageFeature, FeatureNode>
+) {
+    /** Applies the specified function [f] to a freshly created [FeaturesView] and returns its state */
+    fun apply(f: FeaturesView.() -> Unit): Map<PackageFeature, FeatureState> =
+        FeaturesView().apply(f).state
+
+    /** Applies the specified function [f] to a freshly created [FeaturesView] and returns its flatten state */
+    fun applyFlat(f: FeaturesView.() -> Unit): Map<String, Map<String, FeatureState>> =
+        FeaturesView().apply(f).stateFlat
+
+    /** Mutable view of a [FeatureGraph] */
+    inner class FeaturesView {
+        val state: MutableMap<PackageFeature, FeatureState> = hashMapOf()
+
+        val stateFlat: MutableMap<String, MutableMap<String, FeatureState>>
+            get() {
+                val map: MutableMap<String, MutableMap<String, FeatureState>> = hashMapOf()
+                for ((feature, state) in state) {
+                    map.getOrPut(feature.pkg.rootDirectory.toString()) { hashMapOf() }[feature.name] = state
+                }
+                return map
+            }
+
+        init {
+            for (feature in featureToNode.keys) {
+                state[feature] = FeatureState.Disabled
+            }
+        }
+
+        fun enable(feature: PackageFeature) {
+            val node = featureToNode[feature] ?: return
+            enableFeatureTransitively(node)
+        }
+
+        fun disable(feature: PackageFeature) {
+            val node = featureToNode[feature] ?: return
+            disableFeatureTransitively(node)
+        }
+
+        private fun enableFeatureTransitively(node: FeatureNode) {
+            state[node.data] = FeatureState.Enabled
+
+            for (edge in graph.incomingEdges(node)) {
+                val dependency = edge.source
+                enableFeatureTransitively(dependency)
```

Comment:
It might be a good idea to check if the feature is not already enabled:

* to avoid 2^n perf for patahlogical cases
* to avoid infinite loops for cyclic features (I think cargo either already allows cyclic feature, or there's an issue which is accepted for impl)